src/lib.rs

```diff
@@ -1147,8 +1150,128 @@ pub enum Expression {
     ArrayLength(Handle<Expression>),
 }
 
-/// A code block is just a vector of statements.
-pub type Block = Vec<Statement>;
+/// A code block is a vector of statements, with maybe a vector of spans.
+#[derive(Debug, Clone, Default)]
+#[cfg_attr(feature = "serialize", derive(serde::Serialize))]
+#[cfg_attr(
+    any(feature = "serialize", feature = "deserialize"),
+    serde(transparent)
+)]
+pub struct Block {
+    body: Vec<Statement>,
+    #[cfg(feature = "span")]
+    #[cfg_attr(any(feature = "serialize", feature = "deserialize"), serde(skip))]
+    span_info: Vec<Span>,
+}
+
+impl Block {
+    pub fn new() -> Self {
+        Self {
+            body: Vec::new(),
+            #[cfg(feature = "span")]
+            span_info: Vec::new(),
+        }
+    }
+
+    pub fn from_vec(body: Vec<Statement>) -> Self {
+        #[cfg(feature = "span")]
+        let span_info = std::iter::repeat(Span::Unknown).take(body.len()).collect();
+        Self {
+            body,
+            #[cfg(feature = "span")]
+            span_info,
+        }
+    }
+
+    pub fn body(&self) -> &Vec<Statement> {
+        &self.body
+    }
+
+    #[allow(unused_variables)]
+    pub fn push(&mut self, end: Statement, span: Span) {
+        self.body.push(end);
+        #[cfg(feature = "span")]
+        self.span_info.push(span);
+    }
+
+    pub fn extend(&mut self, item: Option<(Statement, Span)>) {
+        if let Some((end, span)) = item {
+            self.push(end, span)
+        }
+    }
+
+    pub fn extend_block(&mut self, other: Self) {
+        #[cfg(feature = "span")]
+        self.span_info.extend(other.span_info);
+        self.body.extend(other.body);
+    }
+
+    pub fn cull<R: RangeBounds<usize> + Clone>(&mut self, range: R) {
+        #[cfg(feature = "span")]
+        self.span_info.drain(range.clone());
+        self.body.drain(range);
+    }
+
+    pub fn splice<R: RangeBounds<usize> + Clone>(&mut self, range: R, other: Self) {
+        #[cfg(feature = "span")]
+        self.span_info
+            .splice(range.clone(), other.span_info.into_iter());
+        self.body.splice(range, other.body.into_iter());
+    }
+
+    pub fn iter(&self) -> std::slice::Iter<Statement> {
+        self.body.iter()
+    }
+
+    pub fn iter_mut(&mut self) -> impl Iterator<Item = (&mut Statement, Option<&mut Span>)> {
```

Comment:
this is a bit weird. There is no symmetry between `iter()` and `iter_mut()` in terms of returned type?

Reply:
Renamed iter_mut to a different beast, and removed iter altogether since it's provided by Deref.